### PR TITLE
Added MMS:CLI, MMS:Config; updated tests

### PR DESCRIPTION
**Pull request by kris-lab**
See meta issue #8

### CODE REVIEW

**Comment by kris-lab** [comment]
@tomaszdurka please review

This is huge PR which contains a lot of changes:
* introduced CLI class (bin/mms-api reduced to minimum)
* passing `data` to constructor instead of `ids`
* introduced tests
* introduced CONFIG class (`~/.mms-api`)

I could give you a guide if you prefer?

**Comment by tomaszdurka** [comment]
Started to look into it.

File: README.md [hunk 1]
```diff
@@ -16,11 +16,20 @@ mms-api is a tool for accessing MMS API
...
+  List:
+      groups | hosts | clusters | snapshots | alerts | restorejobs
+  Create:
+      restorejob-create <now|timestamp|snapshot-id>
+      alert-ack <group-id> <all|alert-id> <forever|timestamp>
+
+Arguments:
+
+    ID: string[24]
+    TIMESTAMP: `YYYY-MM-DD H:M:S`
```
Comment (tomaszdurka) [inline README.md]: Is this autogenerated help message?

**Comment by tomaszdurka** [comment]
Got this error
```
bundle exec bin/mms-api
Error: undefined method `[]' for nil:NilClass
Available actions: GROUPS, HOSTS, CLUSTERS, SNAPSHOTS, ALERTS, RESTOREJOBS, RESTOREJOB-CREATE, ALERT-ACK
```

File: bin/mms-api [hunk 2]
```diff
@@ -3,149 +3,5 @@
...
-      end
-  end
-
-  puts Terminal::Table.new :title => action.upcase, :headings => (heading.nil? ? [] : heading), :rows => rows
-
-rescue => e
-  puts "Error: `#{e.message}`"
-  exit 1
-end
+MMS::CLI.parse_options
```
Comment (tomaszdurka) [inline bin/mms-api]: I would rather pass ARGV here than make it default for `parse_options` method.

**Comment by kris-lab** [comment]
I believe this is due to non-existent config file. You could create `~/.mms-api` with content:
```
username=hello@example.com
apikey=xxxxxxxxxxxxxxxxxxxx
action=hosts
group_id=53a29fcfc84d93219d181511
cluster_id=53a47d00e4b055d8db1e38b3
```

Hmm please try `bundle exec bin/mms-api -u username -k apikey hosts`

But in general I agree that this is bug. Requires better error handling.

**Comment by tomaszdurka** [comment]
Oh I've created config i have 4 values you mentioned (except action).
This happens as there is no default action?

**Comment by kris-lab** [comment]
If there is no action I have:
````
$ mms-api 
Error: Unknown action 
Available actions: GROUPS, HOSTS, CLUSTERS, SNAPSHOTS, RESTOREJOBS, RESTOREJOBS-CREATE
mms-api is a tool for accessing MMS API

Usage:

	mms-api command [options]

Commands:

	groups | hosts | clusters | snapshots | restorejobs | restorejobs-create

Options:

    -u, --username <string>          MMS user
    -k, --apikey <string>            MMS api-key
    -a, --apiurl <string>            MMS api url. Full url including version: https://mms.mydomain.tld/api/public/v1.0
    -n, --name <string>              Filter for resource name using regexp
    -l, --limit <int>                Limit for result items
    -v, --version                    Version
    -h, --help                       Show this help
```
which is what make sense?

File: lib/mms/agent.rb [hunk 3]
```diff
@@ -30,9 +35,9 @@ def set_apiurl(apiurl)
     def groups
       group_list = []
       MMS::Client.instance.get('/groups').each do |group|
-        group_list.push MMS::Resource::Group.new group['id'], group
+        group_list.push MMS::Resource::Group.new(group)
       end
-      group_list
+      group_list.select { |group| group.id == @default_group or @default_group.nil? }
```
Comment (tomaszdurka) [inline lib/mms/agent.rb]: First checking if nil would be probably more efficient and understandable?

File: lib/mms/agent.rb [hunk 4]
```diff
@@ -48,7 +53,7 @@ def clusters
       groups.each do |group|
         cluster_list.concat group.clusters
       end
-      cluster_list
+      cluster_list.select { |cluster| cluster.id == @default_cluster or @default_cluster.nil? }
```
Comment (tomaszdurka) [inline lib/mms/agent.rb]: Same as above.

**Comment by tomaszdurka** [comment]
As discussed with @kris-lab 

- Make Client non-singleton
- Pass Client to each resource while instantiating
- Make Config non-singleton
- Pass Config to Agent while creating Agent
- Consider keeping or disabling default_group/default_cluster (would need to change cli framework - no magic send)

File: lib/mms/agent.rb (hunk 3 above)
Comment (tomaszdurka) [inline lib/mms/agent.rb]: Maybe rather `group_list.select { |group| group.id == @default_group} unless @default_group.nil?`

**Comment by kris-lab** [comment]
@tomaszdurka could you have a look on first part of refactoring. Basically I got rid of singletons as we discussed + added client instance to agent + resource.

IMO it goes to right direction. So please have a look and let me know when could we have a call?

**Comment by kris-lab** [comment]
@tomaszdurka also it would be good to pair on it? so then we can close this topic pretty much in 1 day:)?

**Comment by tomaszdurka** [comment]
As agreed we pair on it today 12 pm.

File: lib/mms/agent.rb [hunk 5]
```diff
@@ -2,8 +2,13 @@ module MMS
 
   class Agent
 
-    def initialize(username, apikey)
-      MMS::Client.instance.auth_setup(username, apikey)
+    attr_accessor :client
+
+    def initialize(config = nil, client = nil)
```
Comment (tomaszdurka) [inline lib/mms/agent.rb]: Pass only `config` or only `client`.

File: lib/mms/agent.rb [hunk 6]
```diff
@@ -19,20 +24,21 @@ def set_apiurl(apiurl)
...
     end
 
     def groups
-      group_list = []
-      MMS::Client.instance.get('/groups').each do |group|
-        group_list.push MMS::Resource::Group.new group['id'], group
-      end
-      group_list
+      group_list = MMS::Resource::Group.findGroups(@client)
+      group_list.select { |group| group.id == @client.config.group_id or @client.config.group_id.nil? }
```
Comment (tomaszdurka) [inline lib/mms/agent.rb]: `group_list.select { |group| group.id == @client.config.group_id) unless @client.config.group_id.nil?`

File: lib/mms/agent.rb [hunk 7]
```diff
@@ -48,7 +54,7 @@ def clusters
       groups.each do |group|
         cluster_list.concat group.clusters
       end
-      cluster_list
+      cluster_list.select { |cluster| cluster.id == @client.config.cluster_id or @client.config.cluster_id.nil? }
```
Comment (tomaszdurka) [inline lib/mms/agent.rb]:
```ruby
cluster_list.select { |cluster| @config.groups.include? cluster.id }
```

**Comment by tomaszdurka** [comment]
- Move filtering of groups and clusters out of agent/client
- *Simplify* and structure Cli
  - Use one class or at least one method per command/action (look into Clamp and its subcommands)
  - Perform filtering, validation in actions themeselves (do not use loop/hash checking)

**Comment by kris-lab** [comment]
@tomaszdurka could you have a look; indeed, it looks much better with `clump`!

I think we could really push this topic forward soon as you are in Basel?

File: lib/mms/resource/cluster.rb [hunk 8]
```diff
@@ -12,39 +14,109 @@ class Resource::Cluster < Resource
...
+      while tries > 0
+        begin
+          restore_jobs = restorejobs
+          tries = 0
+        rescue Exception => e
+          tries-=1;
+          raise MMS::ResourceError.new(e.message, self) if tries < 1
+
+          puts e.message
+          puts 'Sleeping for 5 seconds. Trying again...'
```
Comment (tomaszdurka) [inline lib/mms/resource/cluster.rb]: use `STDERR.puts`?

**Comment by tomaszdurka** [comment]
Looks good to me.

As agreed I will still improve readme and code documentation in #14.

**Pull request merged**